coderd/workspaceupdates_test.go

```diff
@@ -0,0 +1,345 @@
+package coderd_test
+
+import (
+	"context"
+	"encoding/json"
+	"slices"
+	"strings"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+
+	"cdr.dev/slog/sloggers/slogtest"
+	"github.com/coder/coder/v2/coderd"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+	"github.com/coder/coder/v2/coderd/database/pubsub"
+	"github.com/coder/coder/v2/coderd/rbac"
+	"github.com/coder/coder/v2/coderd/rbac/policy"
+	"github.com/coder/coder/v2/coderd/wspubsub"
+	"github.com/coder/coder/v2/tailnet"
+	"github.com/coder/coder/v2/tailnet/proto"
+	"github.com/coder/coder/v2/testutil"
+)
+
+func TestWorkspaceUpdates(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
```

Comment:
You should use contexts that timeout after `testutil.WaitShort`, so that this won't hang the test suite if stuff fails.

BUT! You need to initialize the context in each subtest after `t.Parallel()` --- not here!

Reply:
I believe I caused a flake by not doing this in another test, I'm very much curious as to why this needs to be done.

`testutil.Context` just calls `t.Cleanup` with the context cancellation function, and `t.Cleanup` runs passed callbacks once the test *and* all it's subtests finish. So, if you had a context in the main test and used it in the subtests, it shouldn't get cancelled early. Does the way `t.Parallel()` work influence it somehow?

Reply:
`testutil.Context` includes a timeout, which starts from the time the test function is called.  But, parallel subtests get run asynchronously, in parallel, so an uncontrolled amount of time passes before the subtests start, at which point, the context might have already expired.

I wrote a blog post about it last year: https://coder.com/blog/go-testing-contexts-and-t-parallel

Reply:
Ahh, I see - thanks Spike!